File: java/revenj-core/src/main/java/org/revenj/database/postgres/converters/BoolConverter.java
package org.revenj.database.postgres.converters;

import org.revenj.database.postgres.PostgresBuffer;
import org.revenj.database.postgres.PostgresWriter;
import org.revenj.database.postgres.PostgresReader;

import java.util.ArrayList;
import java.util.List;

public abstract class BoolConverter {

	private static final char[] TRUE = "true".toCharArray();
	private static final char[] FALSE = "false".toCharArray();

	public static void serializeURI(PostgresBuffer sw, Boolean value) {
		if (value == null) return;
		if (value) {
			sw.addToBuffer(TRUE);
		} else {
			sw.addToBuffer(FALSE);
		}
	}

	public static Boolean parseNullable(PostgresReader reader) {
		int cur = reader.read();
		if (cur == ',' || cur == ')') {
			return null;
		}
		reader.read();
		return cur == 't';
	}

	public static boolean parse(PostgresReader reader) {
		int cur = reader.read();
		if (cur == ',' || cur == ')') {
			return false;
		}
		reader.read();
		return cur == 't';
	}

	public static List<Boolean> parseCollection(PostgresReader reader, int context, boolean allowNulls) {
		int cur = reader.read();
		if (cur == ',' || cur == ')') {
			return null;
		}
		boolean escaped = cur != '{';
		if (escaped) {
			reader.read(context);
		}
		List<Boolean> list = new ArrayList<>();
		cur = reader.peek();
		if (cur == '}') {
			reader.read();
		}
		Boolean defaultValue = allowNulls ? null : false;
		while (cur != -1 && cur != '}') {
			cur = reader.read();
			if (cur == 't') {
				list.add(true);
			} else if (cur == 'f') {
				list.add(false);
			} else {
				reader.read(3);
				list.add(defaultValue);
			}
			cur = reader.read();
		}
		if (escaped) {
			reader.read(context + 1);
		} else {
			reader.read();
		}
		return list;
	}

	public static PostgresTuple toTuple(Boolean value) {
		if (value == null) return null;
		return new BoolTuple(value);
	}

	public static PostgresTuple toTuple(boolean value) {
		return new BoolTuple(value);
	}

	private static class BoolTuple extends PostgresTuple {
		private final char value;

		BoolTuple(boolean value) {
			this.value = value ? 't' : 'f';
		}

		public boolean mustEscapeRecord() {
			return false;
		}

		public boolean mustEscapeArray() {
			return false;
		}

		public void insertRecord(PostgresWriter sw, String escaping, Mapping mappings) {
			sw.write(value);
		}

		public void insertArray(PostgresWriter sw, String escaping, Mapping mappings) {
			sw.write(value);
		}

		public String buildTuple(boolean quote) {
			return quote ? "'" + value + "'" : String.valueOf(value);
		}
	}
}

File: scala/revenj-core/src/main/scala/net/revenj/database/postgres/converters/BoolConverter.scala
package net.revenj.database.postgres.converters

import net.revenj.database.postgres.{PostgresBuffer, PostgresReader, PostgresWriter}

object BoolConverter extends Converter[Boolean] {

  override def serializeURI(sw: PostgresBuffer, value: Boolean): Unit = {
    sw.addToBuffer(if (value) "true" else "false")
  }

  override val dbName = "bool"

  override def default() = false

  private val someTrue = Some(true)
  private val someFalse = Some(false)

  override def parse(reader: PostgresReader, context: Int): Boolean = {
    val cur = reader.read()
    if (cur == ',' || cur == ')') {
      false
    } else {
      reader.read()
      cur == 't'
    }
  }

  override def parseOption(reader: PostgresReader, context: Int): Option[Boolean] = {
    val cur = reader.read()
    if (cur == ',' || cur == ')') {
      None
    } else {
      reader.read()
      if (cur == 't') someTrue
      else someFalse
    }
  }

  override def parseRaw(reader: PostgresReader, start: Int, context: Int): Boolean = {
    reader.read()
    start == 't'
  }

  override def parseCollectionItem(reader: PostgresReader, context: Int): Boolean = {
    val cur = reader.read()
    if (cur == 't') {
      reader.read()
      true
    } else if (cur == 'f') {
      reader.read()
      false
    } else {
      reader.read(4)
      false
    }
  }

  override def parseNullableCollectionItem(reader: PostgresReader, context: Int): Option[Boolean] = {
    val cur = reader.read()
    if (cur == 't') {
      reader.read()
      someTrue
    } else if (cur == 'f') {
      reader.read()
      someFalse
    } else {
      reader.read(4)
      None
    }
  }

  def toTuple(value: Boolean): PostgresTuple = {
    new BooleanTuple(value)
  }

  private class BooleanTuple(val value: Boolean) extends PostgresTuple {
    val charValue: Char = if (value) 't' else 'f'
    val mustEscapeRecord = false

    val mustEscapeArray = false

    def insertRecord(sw: PostgresWriter, escaping: String, mappings: Option[(PostgresWriter, Char) => Unit]): Unit = {
      sw.write(charValue)
    }

    override def buildTuple(quote: Boolean): String = {
      if (quote) {
        val buf = new Array[Char](3)
        buf(0) = '\''
        buf(1) = charValue
        buf(2) = '\''
        new String(buf, 0, buf.length)
      } else String.valueOf(charValue)
    }
  }

}
